rust/agama-dbus-server/src/locale.rs

```diff
@@ -0,0 +1,166 @@
+                language
+                    .names
+                    .name_for(DEFAULT_LANG)
+                    .context("missing default translation for language")?,
+                territory
+                    .names
+                    .name_for(DEFAULT_LANG)
+                    .context("missing default translation for territory")?,
+            );
+            let localized_ret = (
+                language
+                    .names
+                    .name_for(language.id.as_str())
+                    .context("missing native label for language")?,
+                territory
+                    .names
+                    .name_for(language.id.as_str())
+                    .context("missing native label for territory")?,
+            );
+            res.push((default_ret, localized_ret));
+        }
+
+        Ok(res)
+    }
+
+    #[dbus_interface(property)]
+    fn locales(&self) -> Vec<String> {
+        return self.locales.iter().map(|l| l.clone()).collect();
+    }
+
+    #[dbus_interface(property)]
+    fn set_locales(&mut self, locales: Vec<String>) -> Result<(), zbus::fdo::Error> {
+        // verify that all locales are supported
+        self.locales = locales;
+        Ok(())
+    }
+
+    /* support only keymaps for console for now
+        fn list_x11_keyboards(&self) -> Result<Vec<(String, String)>, Error> {
+            let keyboards = agama_locale_data::get_xkeyboards()?;
+            let ret = keyboards
+                .keyboard.iter()
+                .map(|k| (k.id.clone(), k.description.clone()))
+                .collect();
+            Ok(ret)
+        }
+
+        fn set_x11_keyboard(&mut self, keyboard: &str) {
+            self.keyboard_id = keyboard.to_string();
+        }
+    */
+
+    #[dbus_interface(name="ListVConsoleKeyboards")]
+    fn list_keyboards(&self) -> Result<Vec<String>, Error> {
+        let res = agama_locale_data::get_key_maps()?;
+        Ok(res)
+    }
+
+    #[dbus_interface(property, name="VConsoleKeyboard")]
+    fn keymap(&self) -> &str {
+        return &self.keymap.as_str();
+    }
+
+    #[dbus_interface(property, name="VConsoleKeyboard")]
+    fn set_keymap(&mut self, keyboard: &str) -> Result<(), zbus::fdo::Error> {
+        self.keymap = keyboard.to_string();
+        Ok(())
+    }
+
+    fn list_timezones(&self, locale: &str) -> Result<Vec<(String, String)>, Error> {
+        let timezones = agama_locale_data::get_timezones();
+        let localized =
+            agama_locale_data::get_timezone_parts()?.localize_timezones(locale, &timezones);
+        let ret = timezones.into_iter().zip(localized.into_iter()).collect();
+        Ok(ret)
+    }
+
+    #[dbus_interface(property)]
+    fn timezone(&self) -> &str {
+        return &self.timezone_id.as_str();
+    }
+
+    #[dbus_interface(property)]
+    fn set_timezone(&mut self, timezone: &str) -> Result<(), zbus::fdo::Error> { // NOTE: cannot use crate::Error as property expect this one
```

Comment:
I see, that's fine, it does have a generic [`zbus::fdo::Error::Failed(String)` constructor](https://docs.rs/zbus/3.12.0/zbus/fdo/enum.Error.html#variant.Failed)

Reply:
OK, so I will add some validation code with it, so we can test it

Reply:
You can convert from one type into the other if needed. For instance,

```rust
impl From<MyErrorType> for zbus::fdo::Error {
    fn from(value: MyErrorType) -> zbus::fdo::Error {
        zbus::fdo::Error::Failed(format!("Some error: {}", value.to_string()))
    }
}
```

So when using a `?` it gets automatically converted.